### PR TITLE
feat: replace `map_foreach` function with `foreach` for iterating through iterables (array, map, string)

### PR DESCRIPTION
**Pull request by its-darsh**

### CODE REVIEW

File: src/vm/builtins/Array.zig [hunk 1]
```diff
@@ -60,9 +61,34 @@ fn array_reverse(vm: *VirtualMachine, arguments: []const Code.Value) Code.Value
     return Code.Value.Object.Array.init(vm.allocator, new_array) catch Code.Value{ .none = {} };
 }
 
-fn length(vm: *VirtualMachine, arguments: []const Code.Value) Code.Value {
-    _ = vm;
+fn foreach(vm: *VirtualMachine, arguments: []const Code.Value) Code.Value {
+    if (!((arguments[0] == .object and (arguments[0].object == .array or arguments[0].object == .map)) and (arguments[1] == .object and arguments[1].object == .function))) {
```
Comment (yhyadev) [inline src/vm/builtins/Array.zig:65]: No need for checking beforehand if you used a switch statement

File: src/vm/builtins/Array.zig [hunk 2]
```diff
@@ -60,9 +61,34 @@ fn array_reverse(vm: *VirtualMachine, arguments: []const Code.Value) Code.Value
...
+        return Code.Value{ .none = {} };
+    }
+
+    const callback = arguments[1].object.function;
+
+    if (arguments[0].object == .map) {
+        const map = arguments[0].object.map;
```
Comment (yhyadev) [inline src/vm/builtins/Array.zig:72]: Use a switch statement in here, `switch (arguments[0]) {}` and leave else cases empty until the end you just return none, check `length` function for an example

File: src/vm/builtins/Array.zig [hunk 3]
```diff
@@ -60,9 +61,34 @@ fn array_reverse(vm: *VirtualMachine, arguments: []const Code.Value) Code.Value
...
+        var map_iter = map.inner.iterator();
+
```
Comment (yhyadev) [inline src/vm/builtins/Array.zig:74]: map_entry_iterator

File: src/vm/builtins/Array.zig [hunk 4]
```diff
@@ -60,9 +61,34 @@ fn array_reverse(vm: *VirtualMachine, arguments: []const Code.Value) Code.Value
...
+        while (map_iter.next()) |map_entry| {
```
Comment (yhyadev) [inline src/vm/builtins/Array.zig:75]: map_entry

File: src/vm/builtins/Array.zig [hunk 5]
```diff
@@ -60,9 +61,34 @@ fn array_reverse(vm: *VirtualMachine, arguments: []const Code.Value) Code.Value
...
+            const args = [2]Code.Value{ map_entry.key_ptr.*, map_entry.value_ptr.* };
+            _ = callback.call(vm, &args);
+        }
+        return Code.Value{ .none = {} };
+    }
+    const arr = arguments[0].object.array;
```
Comment (yhyadev) [inline src/vm/builtins/Array.zig:81]: array, not arr, follow code style

File: src/vm/builtins/Array.zig [hunk 6]
```diff
@@ -60,9 +61,34 @@ fn array_reverse(vm: *VirtualMachine, arguments: []const Code.Value) Code.Value
...
+
+    for (0..arr.values.items.len) |i| {
+        const args = [1]Code.Value{arr.values.items[i]};
+        _ = callback.call(vm, &args);
+    }
+
+    return Code.Value{ .none = {} };
+}
 
+fn length(_: *VirtualMachine, arguments: []const Code.Value) Code.Value {
```
Comment (yhyadev) [inline src/vm/builtins/Array.zig:91]: Why did you change it to that?

**Review by yhyadev — CHANGES_REQUESTED** [review]
There is just coding style comments to change, Good job anyways

File: src/vm/builtins/Array.zig (hunk 6 above)
Comment (its-darsh) [reply src/vm/builtins/Array.zig:91]: it looks better now

File: src/vm/builtins/Array.zig (hunk 5 above)
Comment (its-darsh) [reply src/vm/builtins/Array.zig:81]: make senes

File: src/vm/builtins/Array.zig (hunk 4 above)
Comment (its-darsh) [reply src/vm/builtins/Array.zig:75]: what's wrong?

File: src/vm/builtins/Array.zig (hunk 3 above)
Comment (its-darsh) [reply src/vm/builtins/Array.zig:74]: its already self describable, `map_iter` would _probably_ translate to `map_iterator`, we later specify the type of iterator's items (`map_entery`)

File: src/vm/builtins/Array.zig (hunk 4 above)
Comment (its-darsh) [reply src/vm/builtins/Array.zig:75]: my suggestion is to use `map_iter` for the iterator and `entry` as variable names instead

File: src/vm/builtins/Array.zig (hunk 3 above)
Comment (yhyadev) [reply src/vm/builtins/Array.zig:74]: It is just more verbose and less devious, you could just say map.keyIterator or map.valueIterator and it is still map_iter, by asking the reader to read more for context is like sipping into a muddy water and at the end there is milk

**Review by yhyadev — CHANGES_REQUESTED** [review]
Do it for String too <:

**Comment by its-darsh** [comment]
all aboard?

**Review by yhyadev — APPROVED** [review]
Ok, Good Enough
